### ISSUE TITLE
[Lang] Deprecate ti.field(ti.f32), use ti.field(ti.f32, None) instead for advanced layouts

### ISSUE THREAD

**Issue by archibate** [taichi-dev/taichi#1614]
**Concisely describe the proposed feature**
I'd like to deprecate `ti.field(ti.f32)`, use `ti.field(ti.f32, shape=None)` instead for advanced layout.

Why? Clarity matters:
1. People who knows about advanced layouts definitely won't bother with adding a `None`.
2. Beginners who just want to declare a dense field will be confused by `Some variable(s) not placed` and get busy seeking where a `shape=xxx` is missed. It would be great if we'd raise an `argument 'shape' not specified` error for them.

**Describe the solution you'd like (if any)**
Since `ti.field` is a newly-added API, we are OK to tweak it as much as we want before v0.7.0 finally release, we'd **only do this on `ti.field`, not `ti.var`**, both ensuring backward-compatib and pushing refactoring forward.
Simply change `def field(dt, shape=None)` -> `def field(dt, shape)` should be OK.

**Additional comments**
@k-ye @yuanming-hu WDYT about this change?

**Comment by k-ye** [comment]
> It would be great if we'd raise an argument 'shape' not specified error for them.

+1 on this.

In addition to the verbosity, having to write `shape=None` when using advanced layout also confuses me, although I guess it reduces the confusion if this is more explicit, e.g. `shape=ti._known_after_place` or something (which makes things even more verbose).

Can we just improve the error message first and decide whether `shape` is needed in all cases later?

**Comment by k-ye** [comment]
> Too verbose to me.

If this defaults to such a marker (instead of `None`, for more clarity), then no need to write out `shape` where it's not needed.

**Comment by archibate** [comment]
But my reason was:
This can confuse beginners.
As `import this` said: explicit is always better than implicit.

**Comment by yuanming-hu** [comment]
Thanks for proposing this. I feel like enforcing `shape=None` can be a little confusing for people who are using advanced data layouts. 

> Can we just improve the error message first and decide whether `shape` is needed in all cases later?

I agree with this - we should just improve the error message for now. We can record the stack backtrace of `ti.field/ti.Vector/Matrix/field` in the `SNode`s, and when printing "Some variable(s) not placed", print the backtrace as well so that users can easily pinpoint which field is not placed.

Btw, let's change ""Some variable(s) not placed" to ""Some fields(s) not placed".

**Comment by archibate** [comment]
> We can record the stack backtrace of

That sounds overkill to me, given that most beginners won't have any idea about advanced layouts. It won't be a big deal for 'advanced people' to write `ti.field(ti.f32, None)`, isn't it?
~Or, how about `ti.ad_field(ti.f32)` if you are worrying the line length?~

**Comment by archibate** [comment]
So, shall I merge #1615 or not?
Let's switch to the high-demanding API `ti.field(ti.f32, None)` for now. We can easily switch back to the low-demanding API `ti.field(ti.f32)` when we want.
But if we keep the low-demanding API `ti.field(ti.f32)`. Then it will be hard to switch to the high-demanding one later.

**Comment by yuanming-hu** [comment]
I don't think we should force users to use `shape=None` when using advanced layouts. If the goal is to help beginners to identify which fields are not placed, we should use the stack backtrace approach I mentioned above.

**Comment by archibate** [comment]
I'd suggest to default `dtype=ti.f32` for making advanced people happy :)
